lib/elixir/test/elixir/module/checker_test.exs

```diff
@@ -0,0 +1,486 @@
+        def b(a, b), do: a or b
+      end
+      """
+    }
+
+    assert_no_warnings(files)
+  end
+
+  test "hints exclude deprecated functions" do
+    files = %{
+      "a.ex" => """
+      defmodule A do
+        def to_charlist(a), do: a
+
+        @deprecated "Use String.to_charlist/1 instead"
+        def to_char_list(a), do: a
+
+        def c(a), do: A.to_list(a)
+      end
+      """
+    }
+
+    warning = """
+    warning: A.to_list/1 is undefined or private. Did you mean one of:
+
+          * to_charlist/1
+
+      a.ex:7: A.c/1
+
+    """
+
+    assert_warnings(files, warning)
+  end
+
+  test "imports" do
+    files = %{
+      "a.ex" => """
+      defmodule A do
+        import Record
+
+        def a(a, b), do: extract(a, b)
+        def b(arg), do: is_record(arg)
+      end
+      """
+    }
+
+    assert_no_warnings(files)
+  end
+
+  test "requires" do
+    files = %{
+      "a.ex" => """
+      defmodule A do
+        require Integer
+
+        def a(a), do: Integer.is_even(a)
+      end
+      """
+    }
+
+    assert_no_warnings(files)
+  end
+
+  test "excludes no_warn_undefined" do
+    files = %{
+      "a.ex" => """
+      defmodule A do
+        @compile {:no_warn_undefined, [MissingModule, {MissingModule2, :func, 2}]}
+        @compile {:no_warn_undefined, {B, :func, 2}}
+
+        def a, do: MissingModule.func(1)
+        def b, do: MissingModule2.func(1, 2)
+        def c, do: MissingModule2.func(1)
+        def d, do: MissingModule3.func(1, 2)
+        def e, do: B.func(1)
+        def f, do: B.func(1, 2)
+        def g, do: B.func(1, 2, 3)
+      end
+      """,
+      "b.ex" => """
+      defmodule B do
+        def func(_), do: :ok
+      end
+      """
+    }
+
+    warning = """
+    warning: MissingModule2.func/1 is undefined (module MissingModule2 is not available or is yet to be defined)
+      a.ex:7: A.c/0
+
+    warning: MissingModule3.func/2 is undefined (module MissingModule3 is not available or is yet to be defined)
+      a.ex:8: A.d/0
+
+    warning: B.func/3 is undefined or private. Did you mean one of:
+
+          * func/1
+
+      a.ex:11: A.g/0
+
+    """
+
+    assert_warnings(files, warning)
+  end
+
+  defp assert_warnings(files, expected) do
+    in_tmp(fn ->
+      assert capture_compile(files) == expected
+    end)
+  end
+
+  defp assert_no_warnings(files) do
+    in_tmp(fn ->
+      assert capture_compile(files) == ""
+    end)
+  end
+
+  defp capture_compile(files) do
+    files = generate_files(files)
+
+    capture_io(:stderr, fn ->
+      {:ok, modules, _warnings} = Kernel.ParallelCompiler.compile(files)
+
+      Enum.each(modules, fn module ->
+        :code.purge(module)
+        :code.delete(module)
+      end)
+    end)
+  end
+
+  defp generate_files(files) do
+    for {file, contents} <- files do
+      File.write!(file, contents)
+      file
+    end
+  end
+
+  defp in_tmp(fun) do
+    path = PathHelpers.tmp_path(Integer.to_string(System.unique_integer([:positive])))
```

Comment:
```suggestion
    path = PathHelpers.tmp_path("checker")
```

We `rm_rf` the path anyway, so it is fine to be the same location. Otherwise we do it based on the test name.

Reply:
Thanks, this was from back when I had `async: true`.